### PR TITLE
[YUNIKORN-2697] Improve usergroup funtion's test coverage

### PR DESCRIPTION
**Pull request by SP12893678**
### What is this PR for?
Improve usergroup funtion's test coverage
- usergroup.go
- usergroup_test_resolver.go

### What type of PR is it?
* [x] - Test

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2697

### CODE REVIEW

**Comment by SP12893678** [comment]
PTAL, In `TestIntervalCacheCleanUp` func, I use sleep method to trigger the `run` func `case <-time.After(c.interval):` case
But it will encount the **race detected during execution of test** problem.
I have no idea to solve it.

**Comment by pbacsko** [comment]
> PTAL, In `TestIntervalCacheCleanUp` func, I use sleep method to trigger the `run` func `case <-time.After(c.interval):` case But it will encount the **race detected during execution of test** problem. I have no idea to solve it.

I will check in more detail.

**Comment by pbacsko** [comment]
@SP12893678 you cannot access certain variables without locking.

For example line 286:
```
ug = testCache.ugs["unknown"]
```
this should be:
```
ug = testCache.GetUserGroup("unknown")
```

Another example from line 291:
```
	// expire the failed lookup
	ug.resolved -= 2 * negcache
```
this should be protected by the instance lock:
```
	// expire the failed lookup
	testCache.lock.Lock()
	ug.resolved -= 2 * negcache
	testCache.lock.Unlock()
```

You can compare the two call stacks and check what kind of mutexes are involved. In one case, we clearly don't acquire any mutex.

**Review by pbacsko — CHANGES_REQUESTED** [review]
Avoid direct access to `UserGroupCache.ugs`.

Obtain/release the instance mutex when modifying `UserGroup.resolved`.

**Comment by SP12893678** [comment]
> Avoid direct access to `UserGroupCache.ugs`.
> 
> Obtain/release the instance mutex when modifying `UserGroup.resolved`.

Thanks for the suggestion!
When i use `ug = testCache.GetUserGroup("unknown")` instead of `ug = testCache.ugs["unknown"]`
 I found that `ug.resolved -= 2 * negcache` does not work.
So the clear cache condition will not be met to delete ug key

I append a few lock control to solve it.
But it may not a pretty way.

**Review by pbacsko — CHANGES_REQUESTED** [review]
Almost. I have a few feedbacks.

Use `assert.XXX()` in the new tests instead of `t.Errorf()` or `t.Fatalf()` to simplify code. Existing tests can stay as they are.

**Comment by SP12893678** [comment]
> Almost. I have a few feedbacks.
> 
> Use `assert.XXX()` in the new tests instead of `t.Errorf()` or `t.Fatalf()` to simplify code. Existing tests can stay as they are.

Thanks for your reminder.
It looks simply~

File: pkg/common/security/usergroup_test.go [hunk 1]
```diff
@@ -19,34 +19,79 @@
...
 	testCache := GetUserGroupCache("test")
 	testCache.resetCache()
 	// test cache should be empty now
-	if len(testCache.ugs) != 0 {
-		t.Fatalf("Cache not empty: %v", testCache.ugs)
-	}
+	assert.Equal(t, 0, testCache.getUGsize(), "Cache is not empty: %v", testCache.getUGmap())
 
 	ug, err := testCache.GetUserGroup("testuser1")
+	testCache.lock.Lock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:94]: This lock is not needed here. I believe it should be moved right before L#106 when you directly access `testCache.ugs`. `GetUserGroup()` obtains a lock to read the map and returns a **copy** of the `UserGroup` type, not a pointer.

File: pkg/common/security/usergroup_test.go [hunk 2]
```diff
@@ -64,27 +109,29 @@ func TestGetUserGroup(t *testing.T) {
...
 	// click over the clock: if we do not get the cached version the new time will differ from the cache update
 	cachedUG.resolved -= 5
+	testCache.lock.Unlock()
+
 	ug, err = testCache.GetUserGroup("testuser1")
+	testCache.lock.Lock()
 	if err != nil || ug.resolved != cachedUG.resolved {
 		t.Errorf("User 'testuser1' not returned from Cache, resolution time differs: %d got %d (err = %v)", ug.resolved, cachedUG.resolved, err)
 	}
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:119]: This lock can be removed. `GetUserGroup()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 3]
```diff
@@ -64,27 +109,29 @@ func TestGetUserGroup(t *testing.T) {
...
 	ug, err := testCache.GetUserGroup("testuser2")
 	if err != nil {
 		t.Error("Lookup should not have failed: testuser2")
 	}
-	if len(testCache.ugs) != 1 {
-		t.Errorf("Cache not updated should have 1 entry %d", len(testCache.ugs))
-	}
+
+	assert.Equal(t, 1, testCache.getUGsize(), "Cache not updated should have 1 entry %d", testCache.getUGmap())
+	testCache.lock.Lock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:134]: This lock/unlock pair can be removed. `GetUserGroup()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 4]
```diff
@@ -93,64 +140,80 @@ func TestBrokenUserGroup(t *testing.T) {
...
+
+	ug, err = testCache.GetUserGroup("testuser5")
+	assert.ErrorContains(t, err, "lookup failed for user: testuser5")
+
+	ug, err = testCache.GetUserGroup("invalid-gid-user")
+	testCache.lock.Lock()
+	assert.ErrorContains(t, err, "lookup failed for user: invalid-gid-user")
+	exceptedGroup := []string{"1_001"}
+	assert.Assert(t, reflect.DeepEqual(ug.Groups, exceptedGroup), fmt.Errorf("group should be: %v, but got: %v", exceptedGroup, ug.Groups))
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:167]: This lock/unlock pair can be removed. `GetUserGroup()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 5]
```diff
@@ -93,64 +140,80 @@ func TestBrokenUserGroup(t *testing.T) {
...
 	ug, err := testCache.GetUserGroup("")
+	testCache.lock.Lock()
 	if err == nil {
 		t.Error("Lookup should have failed: empty user")
 	}
 	// ug is empty everything should be nil..
 	if ug.User != "" || len(ug.Groups) != 0 || ug.resolved != 0 || ug.failed {
 		t.Errorf("UserGroup is not empty: %v", ug)
 	}
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:186]: This lock/unlock pair can be removed. `GetUserGroup()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 6]
```diff
@@ -93,64 +140,80 @@ func TestBrokenUserGroup(t *testing.T) {
...
 	ug, err = testCache.GetUserGroup("unknown")
+	testCache.lock.Lock()
 	if err == nil {
 		t.Error("Lookup should have failed: unknown user")
 	}
 	// ug is partially filled and failed flag is set
 	if ug.User != "unknown" || len(ug.Groups) != 0 || !ug.failed {
 		t.Errorf("UserGroup is not empty: %v", ug)
 	}
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:198]: This lock/unlock pair can be removed. `GetUserGroup()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 7]
```diff
@@ -93,64 +140,80 @@ func TestBrokenUserGroup(t *testing.T) {
...
 	ug, err = testCache.GetUserGroup("unknown")
+	testCache.lock.Lock()
 	if err == nil {
 		t.Error("Lookup should have failed: unknown user")
 	}
 	// ug is partially filled and failed flag is set: error message should show that the cache was returned
 	if err != nil && !strings.Contains(err.Error(), "cached data returned") {
 		t.Errorf("UserGroup not returned from Cache: %v, error: %v", ug, err)
 	}
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:209]: This lock/unlock pair can be removed. `GetUserGroup()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 8]
```diff
@@ -162,38 +225,78 @@ func TestCacheCleanUp(t *testing.T) {
...
+
+func TestIntervalCacheCleanUp(t *testing.T) {
+	testCache := GetUserGroupCache("test")
+	testCache.resetCache()
+	// test cache should be empty now
+	assert.Equal(t, 0, testCache.getUGsize(), "Cache is not empty: %v", testCache.getUGmap())
+
+	// resolve an existing user
+	_, err := testCache.GetUserGroup("testuser1")
+	assert.NilError(t, err, "Lookup should not have failed: testuser1 user")
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:263]: Small improvement, you can check the `failed` flag here:

```
user1Ug, err := testCache.GetUserGroup("testuser1")
assert.NilError(t, err, "Lookup should not have failed: testuser1 user")
assert.Assert(t, !user1Ug.failed, "User 'testuser1' not resolved as a success")
```

File: pkg/common/security/usergroup_test.go [hunk 9]
```diff
@@ -162,38 +225,78 @@ func TestCacheCleanUp(t *testing.T) {
...
+
+	_, err = testCache.GetUserGroup("testuser2")
+	assert.NilError(t, err, "Lookup should not have failed: testuser1 user")
+
+	testCache.lock.Lock()
+	ug := testCache.ugs["testuser1"]
+	assert.Assert(t, !ug.failed, "User 'testuser1' not resolved as a success")
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:271]: This block can be removed, see above.

File: pkg/common/security/usergroup_test.go [hunk 10]
```diff
@@ -162,38 +225,78 @@ func TestCacheCleanUp(t *testing.T) {
...
+
+	// expire the successful lookup
+	testCache.lock.Lock()
+	ug.resolved -= 2 * poscache
+
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:277]: Resolve the `UserGroup` object for "testuser1" here:

```
	// expire the successful lookup
	testCache.lock.Lock()
	ug := testCache.ugs["testuser1"]
	ug.resolved -= 2 * poscache
	testCache.lock.Unlock()
```

This way it's cleaner and more obvious what you're modifying.

File: pkg/common/security/usergroup_test.go [hunk 11]
```diff
@@ -208,9 +311,11 @@ func TestConvertUGI(t *testing.T) {
 	if err != nil {
 		t.Errorf("known user, no groups, convert should not have failed: %v", err)
 	}
+	testCache.lock.Lock()
 	if ug.User != "testuser1" || len(ug.Groups) != 2 || ug.resolved == 0 || ug.failed {
 		t.Errorf("User 'testuser1' not resolved correctly: %v", ug)
 	}
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:318]: This lock/unlock pair can be removed. `ConvertUGI()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 12]
```diff
@@ -231,12 +336,14 @@ func TestConvertUGI(t *testing.T) {
...
 		t.Errorf("unknown user with groups, convert should not have failed: %v", err)
 	}
+	testCache.lock.Lock()
 	if ug.User != "unknown2" || len(ug.Groups) != 1 || ug.resolved == 0 || ug.failed {
 		t.Fatalf("User 'unknown2' not resolved correctly: %v", ug)
 	}
 	if ug.Groups[0] != group {
 		t.Errorf("groups not initialised correctly on convert: expected '%s' got '%s'", group, ug.Groups[0])
 	}
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:346]: This lock/unlock pair can be removed. `ConvertUGI()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 13]
```diff
@@ -251,4 +358,14 @@ func TestConvertUGI(t *testing.T) {
...
 	}
+
+	// try unknown user with empty group when forced
+	ugi.User = "unknown"
+	ugi.Groups = []string{}
+	ug, err = testCache.ConvertUGI(ugi, true)
+	exceptedGroup := []string{common.AnonymousGroup}
+	testCache.lock.Lock()
+	assert.Assert(t, reflect.DeepEqual(ug.Groups, exceptedGroup), "group should be: %v, but got: %v", exceptedGroup, ug.Groups)
+	testCache.lock.Unlock()
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:369]: This lock/unlock pair can be removed. `ConvertUGI()` returns a copy and internally uses a mutex to access the map.

File: pkg/common/security/usergroup_test.go [hunk 14]
```diff
@@ -162,38 +225,78 @@ func TestCacheCleanUp(t *testing.T) {
...
+	testCache.lock.Lock()
+	ug = testCache.ugs["unknown"]
+	assert.Assert(t, ug.failed, "User 'unknown' not resolved as a failure")
+
+	// expire the failed lookup
+	ug.resolved -= 2 * negcache
+	testCache.lock.Unlock()
+
+	// sleep to wait for interval, it will trigger cleanUpCache
+	time.Sleep(testCache.interval + 1*time.Second)
```
Comment (pbacsko) [inline pkg/common/security/usergroup_test.go:290]: Nit: just `time.Second` instead of `1*time.Second`.

**Review by pbacsko — CHANGES_REQUESTED** [review]
Ok, I did a deeper review of the locking changes.

Quite a few lock/unlock pair can be removed. `ConvertUGI()` and `GetUserGroup()` both return a copy of existing `UserGroup`, so no need to obtain a lock.

Please see my suggested changes
https://github.com/pbacsko/incubator-yunikorn-core/commit/282ab47d9046fd4e4e6e48b0560154e987f030b1

I ran the tests with `-race` multiple times and there were no problems.

**Comment by SP12893678** [comment]
> Ok, I did a deeper review of the locking changes.
> 
> Quite a few lock/unlock pair can be removed. `ConvertUGI()` and `GetUserGroup()` both return a copy of existing `UserGroup`, so no need to obtain a lock.
> 
> Please see my suggested changes [pbacsko@282ab47](https://github.com/pbacsko/incubator-yunikorn-core/commit/282ab47d9046fd4e4e6e48b0560154e987f030b1)
> 
> I ran the tests with `-race` multiple times and there were no problems.

Thanks for having reviewed it and found some unnecessary mutex lock/unlock code.
I refer your suggestion to fix it.
make test` command also passed.

**Review by pbacsko — APPROVED** [review]
+1

**Pull request closed without merging**